sentry-core/src/main/java/io/sentry/core/EnvelopeReader.java

```diff
@@ -0,0 +1,211 @@
+              envelopeBytes, itemHeaderStartOffset, lineBreakIndex - itemHeaderStartOffset);
+
+      if (itemHeader.getLength() <= 0) {
+        throw new IllegalArgumentException(
+            "Item header at index '"
+                + items.size()
+                + "' has an invalid value: '"
+                + itemHeader.getLength()
+                + "'.");
+      }
+
+      payloadEndOffset = lineBreakIndex + itemHeader.getLength() + 1;
+      if (payloadEndOffset > envelopeBytes.length) {
+        throw new IllegalArgumentException(
+            "Invalid length for item at index '"
+                + items.size()
+                + "'. "
+                + "Item is '"
+                + payloadEndOffset
+                + "' bytes. There are '"
+                + envelopeBytes.length
+                + "' in the buffer.");
+      }
+
+      // if 'to' parameter overflows, copyOfRange is happy to pretend nothing happened. Bound need
+      // checking.
+      byte[] envelopeItemBytes =
+          Arrays.copyOfRange(
+              envelopeBytes, lineBreakIndex + 1, payloadEndOffset /* to is exclusive */);
+
+      SentryEnvelopeItem item = new SentryEnvelopeItem(itemHeader, envelopeItemBytes);
+      items.add(item);
+
+      if (payloadEndOffset == envelopeBytes.length) {
+        // End of envelope
+        break;
+      } else if (payloadEndOffset + 1 == envelopeBytes.length) {
+        // Envelope items can be closed with a final line break
+        if (envelopeBytes[payloadEndOffset + 1] == '\n') {
+          break;
+        } else {
+          throw new IllegalArgumentException("Envelope has invalid data following an item.");
+        }
+      }
+
+      itemHeaderStartOffset = payloadEndOffset + 1; // Skip over delimiter
+    } while (true);
+
+    return new SentryEnvelope(header, items);
+  }
+
+  SentryEnvelopeHeader deserializeEnvelopeHeader(byte[] buffer, int offset, int length) {
+    String json = new String(buffer, offset, length, StandardCharsets.UTF_8);
+    return gson.fromJson(json, SentryEnvelopeHeader.class);
+  }
+
+  SentryEnvelopeItemHeader deserializeEnvelopeItemHeader(byte[] buffer, int offset, int length) {
+    String json = new String(buffer, offset, length, StandardCharsets.UTF_8);
+    return gson.fromJson(json, SentryEnvelopeItemHeader.class);
+  }
+
+  private final class SentryEnvelopeHeaderAdapter extends TypeAdapter<SentryEnvelopeHeader> {
+    public void write(JsonWriter out, SentryEnvelopeHeader value) {}
+
+    public SentryEnvelopeHeader read(JsonReader reader) throws IOException {
+      SentryId sentryId = SentryId.EMPTY_ID;
+      String auth = null;
+
+      reader.beginObject();
+      while (reader.hasNext()) {
+
+        switch (reader.nextName()) {
+          case "event_id":
+            sentryId = new SentryId(reader.nextString());
+            break;
+          case "auth":
+            auth = reader.nextString();
+            break;
+          default:
+            reader.skipValue();
```

Comment:
would it not be the case of `unknown` fields?

Reply:
Since there's nothing we can do with it at this point (no envelope endpoint) I'd say no. If we support passing thru envelopes then we'd nerd to keep unknown fields around and adding them to the outgoing envelope